python/ray/workflow/common.py

```diff
@@ -40,18 +39,78 @@ def ensure_ray_initialized():
 
 @dataclass
 class WorkflowRef:
-    """This class represents a dynamic reference of a workflow output.
+    """This class represents a reference of a workflow output.
 
-    See 'step_executor._resolve_dynamic_workflow_refs' for how we handle
-    workflow refs."""
+    The workflow reference could be a **dynamic reference**,
+    **static reference** depending on its content.
+
+    When a reference only contains the step ID, it is a **dynamic reference**.
+    Dynamic reference is used to refer a workflow whose status is unknown.
+    This means the workflow could:
+
+    1. have not executed yet
+    2. have been running
+    3. have failed
+    4. have finished
+
+    Dynamic reference is mostly used for virtual actors in the codebase.
+
+    If we also have the object ref in workflow reference, it is a
+    **unresolved static reference**.Unresolved static reference could be
+    used when you want to return a running workflow from a workflow step.
+    For example, the remaining workflows returned by 'workflow.wait' contain
+    static refs to these pending workflows.
+
+    If we have the output object instance in the workflow reference,
+    it is a **resolved static reference**. Resolved reference represents
+    workflow outputs that are both ready and local. Currently it is used for
+    holding outputs of inplace workflows or to cache outputs loaded from
+    storage during recovery.
+
+    See 'workflow_ref.resolve_workflow_refs' for how we resolve workflow refs.
+    """
     # The ID of the step that produces the output of the workflow.
     step_id: StepID
+    # The object ref that points to the live output of the workflow.
+    output_ref: Optional[ObjectRef] = None
+    # True if we have the output of the workflow.
+    has_output: bool = False
+    # The output of the workflow.
+    output: Optional[Any] = None
```

Comment:
I feel a little bit confused by output_ref and output. Should it be Union[ObjectRef, Any] ?

Do we always have object_ref and output as just caching?